### ISSUE TITLE
Support --grep option to only undebt the files that contain a particular string

### ISSUE THREAD

**Issue by paiweilai** [Yelp/undebt#20]
Most of the time when I use undebt, I grep a particular string to find (then cut/pipe) the filenames I want to run on instead of running on a directory with massive number of files.
Wonder if it worthy to incorporate this feature into undebt?

For example, I normally do:
git grep \<some string\> | cut -d ":" -f 1 | sort | uniq > filenames.txt
cat filenames.txt | xargs -n 1 undebt -p \<my_pattern\> -i

It would be nice to have something like:
undebt -p \<my_pattern\> --grep \<some string\>

**Comment by b5y** [comment]
I can add this enhancement on the next week if no one start doing it.

**Comment by evhub** [comment]
@b5y That would be great! This is definitely a feature we would like to add.

**Comment by ajm188** [comment]
@b5y may be prudent to open a PR once you've started so that others can see that you are working on it (less chance of duplicating work); then just ping us on there when it's ready (ps - @evhub thoughts on this contribution pattern?)

**Comment by b5y** [comment]
@ajm188 Yes, it is good idea. I will open PR soon.
I wanted to ask about assigning this issue to me, but it is possible to do only for contributors, I think.

**Comment by asottile** [comment]
What benefit is this going to have over `git grep -l | xargs undebt ...` or `find ... | xargs undebt ...`?  Seems odd to reinvent the grep wheel.

**Comment by evhub** [comment]
@ajm188 Yeah, makes sense to me. Although you can't open a PR until you actually have at least some changes on your branch.

@asottile Why use `pyparsing` when you can use `regex`? Why use a GUI when you can use the command line? The benefit is simplicity, and especially in this case, it will make it much easier for people to use Undebt in what we think is the right way to use it—that is, grepping for what files to change so Undebt doesn't have to scan every single one.

**Comment by asottile** [comment]
> Why use `pyparsing` when you can use `regex`?
I thought the whole point of pyparsing was you weren't using regular expressions and are actually using a real parser (which is more powerful than regex).

> Why use a GUI when you can use the command line?
Should we add a winforms app to `undebt` :D

> The benefit is simplicity, and especially in this case, it will make it much easier for people to use Undebt in what we think is the right way to use it—that is, grepping for what files to change so Undebt doesn't have to scan every file.

I'd argue that adding complexity isn't really increasing simplicity.  What happens when someone says "well I want perl regular expressions" then you grow a `-P` argument.  What happens when someone says "well I want to filter filenames with globs" then you add some option for that.  What happens when someone says "well I want to filter filenames with regular expressions", "well I only want specifically named folders", etc. etc. etc.  Now your "simple" tool has responsibilities well beyond taking a file and performing operations on it.  Plus there are already tools (`xargs`, `grep`, `git`, etc.) which support all of these operations out of the box and better than you could ever reimplement.

**Comment by evhub** [comment]
@asottile The idea is to make it as easy as possible to do the right thing, and as hard as possible to do the wrong thing. If it's possible to grep to determine what files should be scanned, then that's the right thing to do, since it yields a huge performance increase. But right now, it's much harder than just scanning every file, since you can do that with just one command. The rationale isn't "whenever somebody requests a feature, we add it" but "whenever a feature makes it easier to do the right thing and harder to do the wrong thing." That's why we wouldn't add any of the other examples that you give, but this one is valuable.

**Comment by paiweilai** [comment]
@evhub, @asottile: I agree with that we shouldn't reinvent the wheel and also supporting grep is a pretty useful feature. Instead of making the undebt cmd more complicated, how about we provide a mechanism for people to add hooks to pre-process (or post-process). I am stealing this idea from the pre-commit hook. The goal is to have undebt load the new hooks from a config-file and a script to support the ways people use undebt. In this case, we only need to decide whether a new hook is valuable to be added into this repo, and not worrying about complicating the undebt core (that is, taking a file, a pattern, and process).

**Comment by b5y** [comment]
@paiweilai The idea of calling bash grep was in my mind. I will work with this feature on the next week.
Is there any IRC channel for ```undebt``` project?

**Comment by evhub** [comment]
@b5y Good point! I'm going to create a Gitter chatroom.

**Comment by firba1** [comment]
I'm actually with @asottile on this one. `grep` has a manpage as well as years of examples and expertise. The only thing I'd say is worth doing is grep style syntax of trailing files are inputs (so I don't have to do `git grep -l | sed 's/^/-i /' | xargs undebt ...`, I can just do `git grep -l | xargs undebt ...`.

If the only value we're adding here is not having to pipe to xargs, then the only thing we're doing is reinventing the wheel.

**Comment by evhub** [comment]
@firba1 The problem with that is that if you have `undebt -p pattern` and you tack on `file`, it'll treat it as another pattern file since it comes after `-p`. The way to avoid `sed 's/^/-i /'` is just to do `git grep -l | xargs undebt -p pattern -i`, since `-i`, like `-p`, can accept multiple inputs.

**Comment by ajm188** [comment]
@evhub if that's the main problem we're trying to solve, why not update the cli to have a `--` option, like git subcommands usually do? So then `undebt -p pattern -- file1 file2 ...`

**Comment by firba1** [comment]
This might be worth moving to another issue.

**Comment by evhub** [comment]
@ajm188 The CLI already supports `--`, although it doesn't do anything since we don't have any positional arguments right now. My point is just that `-i` can already do everything that `--` can (just replace `--` with `-i` in your example).

**Comment by asottile** [comment]
Then I'd suggest changing the CLI to instead be:

```
undebt -p pattern [-p pattern ...] FILENAME [FILENAME ...]
```

This is pretty standard for most linting / rewriting tools and certainly helps with `xargs` as above (and would make `undebt` compatible to use as a [pre-commit](http://pre-commit.com) hook)

**Comment by evhub** [comment]
@asottile I think I do actually like that better, although it would break backwards compatibility, so we'd want to bump the version to `0.3.0`.

**Comment by evhub** [comment]
@asottile @firba1 @ajm188 I created #27 for implementing the above change to the CLI.

**Comment by paiweilai** [comment]
After reading all these discussion, I start to agree that instead of making undebt cli more and more complex, probably a better idea is to add a "tips" section in the README/docs to show the best practices of "how to use xargs for parallel processing/directory traversal/grep string".

In this way, we don't have to reinvent wheels, but we still provide guidelines and examples to show how to use undebt efficiently.

**Comment by paiweilai** [comment]
I am going to close this issue as won't fix since we added tips in the doc #31
